README.md
# OctoPrint-Lcd1602

This plug-in allows you to control a 16X2 lcd display (hd44780 connected to port I2C) to display the octoprint status. It is useful for people like me who have a printer without a display.
It indicates on which port the printer is connected, the progress of printing. It also displays the remaining print time (thanks to a simple method).
This plugin has to evolve:
- add options in the interface
- new version for oled displays :)

## Setup

Install via the bundled [Plugin Manager](https://github.com/foosel/OctoPrint/wiki/Plugin:-Plugin-Manager)
or manually using this URL:

    https://github.com/n3bojs4/OctoPrint-Lcd1602/archive/master.zip

**MANUAL INSTALL:** 

clone the repo :

`git clone https://github.com/n3bojs4/OctoPrint-Lcd1602.git  `

install :

`cd OctoPrint-Lcd1602 && python setup.py install`

## Configuration

Nothing to do for the moment :D

CHANGELOG.md
# LCD1602 Octoprint plugin Changelog

## Roadmap

- inprove time calculation left
- add settings in octoprint web interface
- support more devices (oled ?)

## 0.1.1 Bugs corrections
- fix dependencies errors (RPLCD needs smbus2)
- Add gitlab pipeline for avoiding stupid dependencies errors (using fake_rpi)

## 0.1 Initial Release

- Show progress bar with estimated time left.
- Show which printer port is connected
- Eco mode when printer not connected
- Support only PCF8574 expander




octoprint_LCD1602/tests/running_test.sh
#!/bin/bash

if [ $(cat /proc/1/cgroup | grep -c -i docker) -gt 0 ] ; then export LCD1602_DOCKER=1; fi

source /opt/octoprint/venv/bin/activate 
python setup.py install || echo "Error: Cannot install LCD1602 Plugin"

echo "Starting octoprint server"
/opt/octoprint/venv/bin/octoprint serve --iknowwhatimdoing >/tmp/logs &

echo "Checking octoprint is running"
sleep 3
RUNNING=$(pgrep -c octoprint)


if [ $RUNNING -eq 0 ]
then
  echo 'Octoprint failed to start' && exit 1
else
  echo 'Octoprint is started'
fi

echo "Looking for errors on logs"
sleep 10
ERRORS=$(grep -c "^| \!LCD1602 I2c display" /tmp/logs )

if [ $ERRORS -gt 0 ]
then
  echo 'Plugin errors detected, check logs below :'
  grep -A 100 -B 20 'LCD1602' /tmp/logs
  exit 1
else
  echo "Plugin is installed and loaded"
fi


octoprint_LCD1602/tests/raspi-device-mocks/lcd_device.py
#!/usr/bin/env python
# lcd_device: driver code to interface directly with I2C-driven LCD display

''' Origin: Combined Guerreau's 'i2c_device.py' and 'lcddriver.py' into
a single script for importing & controlling. Some minor edits and restructuring,
but the majority of the code base herein is by Guerreau, as stated below.
'''

__author__ = 'Stephane Guerreau'  # Primary author of LCD driver, via GitHub & blog
__credits__ = ['Trevor Allen', ]  # Minor edits & restructuring
# Stephane Guerreau's blog post:
# http://hardware-libre.fr/2014/03/en-raspberry-pi-using-a-4x20-characters-display/
# and associated GitHub repo: https://github.com/CaptainStouf/raspberry_lcd4x20_I2C

## Standard Imports
import time


class lcd(object):
    ## Global variables for LCD object, originally from Guerreau's 'lcddriver.py'
    # LCD Address
    ADDRESS = 0x27

    # commands
    LCD_CLEARDISPLAY = 0x01
    LCD_RETURNHOME = 0x02
    LCD_ENTRYMODESET = 0x04
    LCD_DISPLAYCONTROL = 0x08
    LCD_CURSORSHIFT = 0x10
    LCD_FUNCTIONSET = 0x20
    LCD_SETCGRAMADDR = 0x40
    LCD_SETDDRAMADDR = 0x80

    # flags for display entry mode
    LCD_ENTRYRIGHT = 0x00
    LCD_ENTRYLEFT = 0x02
    LCD_ENTRYSHIFTINCREMENT = 0x01
    LCD_ENTRYSHIFTDECREMENT = 0x00

    # flags for display on/off control
    LCD_DISPLAYON = 0x04
    LCD_DISPLAYOFF = 0x00
    LCD_CURSORON = 0x02
    LCD_CURSOROFF = 0x00
    LCD_BLINKON = 0x01
    LCD_BLINKOFF = 0x00

    # flags for display/cursor shift
    LCD_DISPLAYMOVE = 0x08
    LCD_CURSORMOVE = 0x00
    LCD_MOVERIGHT = 0x04
    LCD_MOVELEFT = 0x00

    # flags for function set
    LCD_8BITMODE = 0x10
    LCD_4BITMODE = 0x00
    LCD_2LINE = 0x08
    LCD_1LINE = 0x00
    LCD_5x10DOTS = 0x04
    LCD_5x8DOTS = 0x00

    # flags for backlight control
    LCD_BACKLIGHT = 0x08
    LCD_NOBACKLIGHT = 0x00

    En = 0b00000100 # Enable bit
    Rw = 0b00000010 # Read/Write bit
    Rs = 0b00000001 # Register select bit

    ## Modified to RECEIVE the bus upon instantiation
    def __init__(self, bus, addr=0x27, port=1):
        # Originally from i2c_lib.py
        self.addr = addr
        # self.bus = smbus.SMBus(port)
        self.bus = bus
        # self.lcd_device = i2c_lib.i2c_device(ADDRESS)

        self.lcd_write(0x03)
        self.lcd_write(0x03)
        self.lcd_write(0x03)
        self.lcd_write(0x02)

        self.lcd_write(self.LCD_FUNCTIONSET | self.LCD_2LINE | self.LCD_5x8DOTS | self.LCD_4BITMODE)
        self.lcd_write(self.LCD_DISPLAYCONTROL | self.LCD_DISPLAYON)
        self.lcd_write(self.LCD_CLEARDISPLAY)
        self.lcd_write(self.LCD_ENTRYMODESET | self.LCD_ENTRYLEFT)
        time.sleep(0.2)

    def write_cmd(self, cmd):
        # Write a single command (originally from i2c_lib.py)
        self.bus.write_byte(self.addr, cmd)
        time.sleep(0.0001)

    def write_cmd_arg(self, cmd, data):
        # Write a command and argument (originally from i2c_lib.py)
        self.bus.write_byte_data(self.addr, cmd, data)
        time.sleep(0.0001)

    def write_block_data(self, cmd, data):
        # Write a block of data (originally from i2c_lib.py)
        self.bus.write_block_data(self.addr, cmd, data)
        time.sleep(0.0001)

    def read(self):
        # Read a single byte (originally from i2c_lib.py)
        return self.bus.read_byte(self.addr)

    def read_data(self, cmd):
        # Read (originally from i2c_lib.py)
        return self.bus.read_byte_data(self.addr, cmd)

    def read_block_data(self, cmd):
        # Read a block of data (originally from i2c_lib.py)
        return self.bus.read_block_data(self.addr, cmd)

    def lcd_strobe(self, data):
        # clocks EN to latch command
        # self.lcd_device.write_cmd(data | En | LCD_BACKLIGHT)
        self.write_cmd(data | self.En | self.LCD_BACKLIGHT)
        time.sleep(.0005)
        # self.lcd_device.write_cmd(((data & ~En) | LCD_BACKLIGHT))
        self.write_cmd(((data & ~self.En) | self.LCD_BACKLIGHT))
        time.sleep(.0001)

    def lcd_write_four_bits(self, data):
        # self.lcd_device.write_cmd(data | LCD_BACKLIGHT)
        self.write_cmd(data | self.LCD_BACKLIGHT)
        self.lcd_strobe(data)

    def lcd_write(self, cmd, mode=0):
        # write a command to lcd
        self.lcd_write_four_bits(mode | (cmd & 0xF0))
        self.lcd_write_four_bits(mode | ((cmd << 4) & 0xF0))

    def lcd_display_string(self, string, line):
        ### String & Line validation applied specifically for SainSmart 20x4 LCD,
        #   which has 4 lines of 20 characters each.
        try:
            line = int(line)
        except ValueError:
            raise ValueError('Line number must be an integer.')
        if line < 1 or line > 4:
            raise ValueError('Line number must be between 1 and 4.')
        # put string function
        if line == 1:
            self.lcd_write(0x80)
        if line == 2:
            self.lcd_write(0xC0)
        if line == 3:
            self.lcd_write(0x94)
        if line == 4:
            self.lcd_write(0xD4)

        for char in string:
            self.lcd_write(ord(char), self.Rs)

    def lcd_clear(self):
        # clear lcd and set to home
        self.lcd_write(self.LCD_CLEARDISPLAY)
        self.lcd_write(self.LCD_RETURNHOME)



octoprint_LCD1602/tests/raspi-device-mocks/rpidevmocks.py
#!/usr/bin/env python
# mocks.py: Mock objects for low-level peripheral/hardware-related modules

''' To-Do:
 - add iterable support channels passed to GPIO
 - add GPIO pull-up/pull-down support
 - add event detection support
'''


__author__ = 'Trevor Allen'

# Standard Library
import random
import time
import math

## RPi.GPIO pins (from http://elinux.org/RPi_Low-level_peripherals):
'''
    RPi A+ & B+ GPIO: J8 40-pin header
    --------------------------------
            +3V3 1  2   +5V
    GPIO2   SDA1 3  4   +5V
    GPIO3   SCL1 5  6   GND
    GPIO4   GCLK 7  8   TXD0  GPIO14
             GND 9  10  RXD0  GPIO15
    GPIO17  GEN0 11 12  GEN1  GPIO18
    GPIO27  GEN2 13 14  GND
    GPIO22  GEN3 15 16  GEN4  GPIO23
            +3V3 17 18  GEN5  GPIO24
    GPIO10  MOSI 19 20  GND
    GPIO9   MISO 21 22  GEN6  GPIO25
    GPIO11  SCLK 23 24  CE0_N GPIO8
             GND 25 26  CE1_N GPIO7
    EEPROM ID_SD 27 28  ID_SC EEPROM
    GPIO5        29 30  GND
    GPIO6        31 32        GPIO12
    GPIO13       33 34  GND
    GPIO19       35 36        GPIO16
    GPIO26       37 38        GPIO20
             GND 39 40        GPIO21
    --------------------------------
    '''


class MockSMBus(object):
    ''' Mock of smbus.SMBus() class '''
    mpu_adr = 0x68
    lcd_adr = 0x27
    pwr_mgt_1 = 0x6b
    ## HIGH byte registers for MPU-6050, value+1 for LOW registers
    accel_x = 0x3b
    accel_y = 0x3d
    accel_z = 0x3f
    gyro_x = 0x43
    gyro_y = 0x45
    gyro_z = 0x47
    temp = 0x41

    def __init__(self, bus_no):
        self.bus_no = bus_no

    def __repr__(self):
        return "<Mock: smbus.SMBus>"

    def write_byte(self, addr, cmd):
        if addr not in (self.mpu_adr, self.lcd_adr):
            errmsg = 'Address argument not valid: %s' % hex(addr)
            raise ValueError(errmsg)
        return None

    def write_byte_data(self, addr, cmd, zero=0):
        ''' orig: sensor.Sensor.initialize() '''
        # res = bus.write_byte_data(0x68, 0x6b, 0)  # res = None
        if addr not in (self.mpu_adr, self.lcd_adr):
            errmsg = 'Address argument not valid: %s' % hex(addr)
            raise ValueError(errmsg)
        ## This is specific to MPU
        if cmd not in (self.mpu_adr, self.lcd_adr, self.pwr_mgt_1, self.accel_x, self.accel_y,
                    self.accel_z, self.gyro_x, self.gyro_y, self.gyro_z, self.temp):
            raise ValueError('Command invalid: %s' % hex(cmd))
        return None

    def write_block_data(self, addr, cmd, data):
        # I don't think this is actually used, just available in lcd_device.py
        if addr not in (self.mpu_adr, self.lcd_adr):
            errmsg = 'Address argument not valid: %s' % hex(addr)
            raise ValueError(errmsg)
        return None

    def read_byte_data(self, addr, cmd):
        ''' orig: sensor.Sensor.read_word() '''
        # addr is mpu-addr (0x68); cmd is HIGH register: accel-x (0x3b), accel-y (0x3d),
        # accel-z (0x3f), gyro-x (0x43), gyro-y (0x45), gyro-z (0x47), temp (0x41)
        results = { self.accel_x : 246, self.accel_x+1 : 88,
                    self.accel_y : 4, self.accel_y+1 : 92,
                    self.accel_z : 57, self.accel_z+1 : 104,
                    self.gyro_x : 6, self.gyro_x+1 : 42,
                    self.gyro_y : 0, self.gyro_y+1 : 184,
                    self.gyro_z : 255, self.gyro_z+1 : 216,
                    self.temp : 240, self.temp+1 : 16,
                    self.lcd_adr : 0
                    }
        if addr not in (self.mpu_adr, self.lcd_adr):
            errmsg = 'Address argument not valid: %s' % hex(addr)
            raise ValueError(errmsg)
        ## This is specific to MPU
        ### Checking commands restricts to MPU, disallows LCD display,
        #   unless one were to add all the LCD attributes below to this class
        if cmd not in results.keys():
            raise ValueError('Argument %s not a valid command.' % hex(cmd))
        try:
            return results[cmd]
        except KeyError:
            return None

    def read_byte(self, addr):
        # I don't think this is actually used, just available in lcd_device.py
        if addr not in (self.mpu_adr, self.lcd_adr):
            errmsg = 'Address argument not valid: %s' % hex(addr)
            raise ValueError(errmsg)
        return None

    def read_block_data(self, addr, cmd):
        # I don't think this is actually used, just available in lcd_device.py
        if addr not in (self.mpu_adr, self.lcd_adr):
            errmsg = 'Address argument not valid: %s' % hex(addr)
            raise ValueError(errmsg)
        return None



class Mock_smbusModule(object):
    ''' Mock of smbus module, containing SMBus class '''
    ### Purpose here is to allow instantiation of a fake smbus 'library'
    #   like so: smbus = Mock_smbusModule(); bus = smbus.SMBus(1)
    SMBus = MockSMBus


class MockGPIO(object):
    # Map format is <BCM-#> : <BOARD-#>
    bcm_board_map = { 2 : 3,
      3 : 5,   4 : 7,   14 : 8,  15 : 10, 17 : 11,
     18 : 12, 27 : 13,  22 : 15, 23 : 16, 24 : 18,
     10 : 19,  9 : 21,  25 : 22, 11 : 23,  8 : 24,
      7 : 26,  5 : 29,   6 : 31, 12 : 32, 13 : 33,
     19 : 35, 16 : 36,  26 : 37, 20 : 38, 21 : 40}
    RPI_REVISION = 2
    BCM = 11
    BOARD = 10
    OUT = 0
    IN = 1
    LOW = 0
    HIGH = 1
    PUD_OFF = 20
    PUD_DOWN = 21
    PUD_UP = 22
    HARD_PWM = 43
    FALLING = 32
    BOTH = 33
    I2C = 42
    RISING = 31
    SERIAL = 40
    SPI = 41
    UNKNOWN = -1
    VERSION = '0.5.11'
    RPI_INFO = {'MANUFACTURER': 'Unknown', 'P1_REVISION': 3, 'PROCESSOR': 'Unknown',
                'RAM': 'Unknown', 'REVISION': '0010', 'TYPE': 'Unknown'}

    gpio_setting = { k:1 for k in bcm_board_map.keys() }

    def __init__(self):
        self.mode = -1
        self.setmode_run = False
        self.setup_run = False
        pass

    def __repr__(self):
        return "<Mock: RPi.GPIO>"

    def setmode(self, mode):
        # mode should be GPIO.BCM or GPIO.BOARD
        if mode not in (self.BCM, self.BOARD):
            raise ValueError('An invalid mode was passed to setmode()')
        self.setmode_run = True
        self.mode = mode
        # Returns nothing
        pass

    def getmode(self):
        # Should return BCM, BOARD, or UNKNOWN
        return self.mode

    def _pin_validate(self, channel):
        ''' For test/mock purposes, to centralize validation checks of pin numbers & values '''
        if channel not in self.bcm_board_map.keys():
            raise ValueError('Channel is invalid on a Raspberry Pi: %s' % str(channel))

    def cleanup(self, channels=None):
        # Resets all to INPUT with no pullup/pulldown and no event detection
        if channels is None:
            channels = self.bcm_board_map.keys()
        elif not hasattr(channels, '__iter__'):
            channels = [channels,]
        for pin in channels:
            self.gpio_setting[pin] = 1
        self.mode = -1
        self.setmode_run = False

    def setup(self, channels, direction, pull_up_down=None, initial=None):
        if not hasattr(channels, '__iter__'):
            channels = [channels, ]
        for channel in channels:
            self._pin_validate(channel)

        if direction not in (self.IN, self.OUT):
            raise ValueError('An invalid direction was passed to setup()')
        if (pull_up_down is not None and
            pull_up_down not in (self.PUD_OFF, self.PUD_UP, self.PUD_DOWN) ):
            raise ValueError('pull_up_down not in pre-defined PUD_OFF/UP/DOWN values')
        self.setup_run = True  # really should do this on a per-channel basis
        self.gpio_setting[channel] = direction
        # Returns nothing
        pass

    def output(self, channel, value):
        if not hasattr(channels, '__iter__'):
            channels = [channels, ]
        for channel in channels:
            self._pin_validate(channel)

        if value not in (self.LOW, self.HIGH):
            raise ValueError('An invalid value was passed to output()')
        if not self.setmode_run:
            raise RuntimeError('Please set pin numbering mode using GPIO.setmode(GPIO.BOARD) or GPIO.setmode(GPIO.BCM)')
        if not self.setup_run:
            raise RuntimeError('The GPIO channel has not been set up as an OUTPUT')
        # Returns nothing
        pass

    def input(self, channels):
        if not hasattr(channels, '__iter__'):
            channels = [channels, ]
        for channel in channels:
            self._pin_validate(channel)

        if not self.setmode_run:
            raise RuntimeError('Please set pin numbering mode using GPIO.setmode(GPIO.BOARD) or GPIO.setmode(GPIO.BCM)')
        if not self.setup_run:
            raise RuntimeError('You must setup() the GPIO channel first')
        # Returns either 0 or 1.
        ### This may need to be customized depending on its intended use, perhaps
        #   by using mock to specify the desired return value in tests. For me
        #   leaving it to return 1 works fine.
        return self.HIGH

    def gpio_function(self, channel):
        self._pin_validate(channel)
        if not self.setmode_run:
            raise RuntimeError('Please set pin numbering mode using GPIO.setmode(GPIO.BOARD) or GPIO.setmode(GPIO.BCM)')
        return  self.gpio_setting[channel]


    ## Following functions are placeholders that need filled it.
    def add_event_callback(self, *args):
        pass

    def add_event_detect(self, *args):
        pass

    def setwarnings(self, *args):
        pass

    def wait_for_edge(self, *args):
        pass

    def event_detected(self, *args):
        pass

    def remove_event_detect(self, *args):
        pass


class MockSPI(object):
    ''' Mock of spi module '''
    ### Designed for use with MFRC522 rfid module by mxgxw, available on GitHub:
    #   https://github.com/mxgxw/MFRC522-python

    spi_reg_names = {
        # Attr name : [register, (spi.xver results)]
        'CRCResultRegM' : [0x21, (0, 255)],
        'CRCResultRegL' : [0x22, (0, 255)],
        'ErrorReg' : [0x06, (0,0)],
        'ControlReg' : [0x0C, (0,16)],
        'TxControlReg' : [0x14, (0,128)],
        'Status2Reg' : [0x08, (0,0)],
        'DivIrqReg' : [0x05, (0,0)],
        'FIFOLevelReg' : [0x0A, (0,0)],
        'CommIrqReg' : [0x04, (0,20)],
        'BitFramingReg' : [0x0D, (0,0)]
    }


    # Attr : [register]
    spi_write_names = {
        'BitFramingReg' : 0x0D,
        'CommandReg' : 0x01,
        'PCD_RESETPHASE' : 0x0F,
        'PCD_CALCCRC' : 0x03,
        'PCD_IDLE' : 0x00,
        'PCD_TRANSCEIVE' : 0x0C,
        'PCD_AUTHENT' : 0x0E,
        'CommIEnReg' : 0x02,
        'ModeReg' : 0x11,
        'TModeReg' : 0x2A,
        'TxAutoReg' : 0x15,
        'TPrescalerReg' : 0x2B,
        'TReloadRegL' : 0x2D,
        'TReloadRegH' : 0x2C
    }



    '''  Need to figure out how to fake these functions from mfrc522.py:

    def CalculateCRC(self, pIndata):
        i = 0
        while i<len(pIndata):
          self.Write_MFRC522(self.FIFODataReg, pIndata[i])
          i = i + 1

    def MFRC522_ToCard(self,command,sendData):
        while(i<len(sendData)):
            self.Write_MFRC522(self.FIFODataReg, sendData[i])
            i = i+1
            self.Write_MFRC522(self.FIFODataReg, sendData[i])
    '''
    def __init__(self):
        raise UserWarning('This mock spi class is not entirely operational yet.')
        self.spi_xfer_vals = {
            0x21 : (0, 255),
            0x22 : (0, 255),
            0x06 : (0,0),
            0x0C : (0,16),
            0x14 : (0,128),
            0x08 : (0,0),
            0x05 : (0,0),
            0x0A : (0,0),
            0x04 : (0,20),
            0x0D : (0,0)
        }
        # Writable Register : Val(s)
        self.spi_write_vals = {
            0x0D : [0x00, 0x07],  # BitFramingReg
            0x01 : [0x0F, 0x03, 0x00, 0x0C, 0x0E],  # CommandReg
            0x02 : [0x00, 0x12, 0x77, 0x80],  # CommIEnReg
            0x11 : [0x3D],  # ModeReg
            0x2A : [0x8D],  # TModeReg
            0x15 : [0x40],  # TxAutoReg
            0x2B : [0x3E],  # TPrescalerReg
            0x2D : [30],  # TReloadRegL
            0x2C : [0],  # TReloadRegH
            0x80 : [0],  # Fallback/Default
            # These following 'masks' may need to be included as "val"s
            #   for all writable registers: 0x03, 0x80, 0x04, 0x08
        }
    def openSPI(self, device='/dev/spidev0.0', speed=1000000):
        # spi.transfer(((addr<<1)&0x7E,val))
        pass

    def closeSPI(self):
        pass

    def transfer(self, stuffs):
        # self.spi.transfer(((addr<<1)&0x7E,val))
        addr, val = stuffs
        ## A very poor attempt to mimic MFRC522's read/write functions:
        if ( (addr not in self.spi_xfer_vals.keys())
            and (addr not in self.spi_write_vals.keys()) ):
            raise KeyError('Addr invalid: %s' % str(addr))
        legit_value = any( [val in vals for vals in self.spi_write_vals.values()] )
        if not legit_value:
            raise ValueError('Value invalid: %s' % hex(val))
        try:
            res = self.spi_xfer_vals[addr]
        except KeyError:
            res = self.spi_write_vals[addr]
        return res




octoprint_LCD1602/__init__.py
# coding=utf-8

"""
  LCD1602 Plugin for Octoprint
"""

from __future__ import absolute_import
from octoprint.printer.estimation import PrintTimeEstimator
import octoprint.plugin
import octoprint.events
from RPLCD.i2c import CharLCD
import time
import datetime
import os
import sys
from fake_rpi import printf
import fake_rpi


class LCD1602Plugin(octoprint.plugin.StartupPlugin,
                    octoprint.plugin.EventHandlerPlugin,
                    octoprint.plugin.ProgressPlugin):

  def __init__(self):
    if (os.getenv('LCD1602_DOCKER')):
      print('We are running in test environnement, no i2c device attached.')
      try:
        print('Loading fake_rpi instead of smbus2')
        sys.modules['smbus2'] = fake_rpi.smbus
        self.mylcd = fake_rpi.smbus.SMBus(1)
      except:
        print('Cannot load fake_rpi !')
    else:
      self.mylcd = CharLCD(i2c_expander='PCF8574', address=0x27, cols=16, rows=2, backlight_enabled=True, charmap='A00')
      
      # create block for progress bar
      self.block = bytearray(b'\xFF\xFF\xFF\xFF\xFF\xFF\xFF')
      self.block.append(255)
      self.mylcd.create_char(1,self.block)
    
    # init vars
    self.start_date = 0

    # create block for progress bar
    #self.mylcd.create_char(1,self.block)

  def JobIsDone(self,lcd):

    # create final anim
    self.birdy = [ '^_-' , '^_^', '-_^' , '^_^', '0_0', '-_-', '^_-', '^_^','@_@','*_*','$_$','<_<','>_>']

    for pos in range(0,13):
      lcd.cursor_pos = (1,pos)
      lcd.write_string(self.birdy[pos])
      time.sleep(0.5)
      lcd.clear()
    lcd.write_string('Job is Done    \,,/(^_^)\,,/')

      
  def on_after_startup(self):
    mylcd = self.mylcd
    self._logger.info("plugin initialized !")

  
  def on_print_progress(self,storage,path,progress):
    mylcd = self.mylcd
    percent = int(progress/6.25)+1
    completed = '\x01'*percent
    mylcd.clear()
    mylcd.write_string('Completed: '+str(progress)+'%')
    mylcd.cursor_pos = (1,0)
    mylcd.write_string(completed)

    if progress==1 :
      self.start_date=time.time()
  
    if progress>10 and progress<100:
      now=time.time()
      elapsed=now-self.start_date
      average=elapsed/(progress-1)
      remaining=int((100-progress)*average)
      remaining=str(datetime.timedelta(seconds=remaining))
      mylcd.cursor_pos = (1,3)
      mylcd.write_string(remaining)

    if progress==100 :
      self.JobIsDone(mylcd)



  def on_event(self,event,payload):
    mylcd = self.mylcd
      
    if event in "Connected":
      mylcd.clear()
      mylcd.write_string('Connected to:')
      mylcd.cursor_pos = (1,0)
      mylcd.write_string(payload["port"])

    if event in "Shutdown":
      mylcd.clear()
      mylcd.write_string('Bye bye ^_^')
      time.sleep(1)
      mylcd._set_backlight_enabled(False)
      mylcd.close()
    
    
    if event in "PrinterStateChanged":
      
      if payload["state_string"] in "Offline":
        mylcd.clear()
        mylcd.write_string('Octoprint is not connected')
        time.sleep(2)
        mylcd.clear()
        mylcd.write_string('saving a polar bear, eco mode ON')
        time.sleep(5)
        mylcd._set_backlight_enabled(False)
      
      if payload["state_string"] in "Operational":
        mylcd._set_backlight_enabled(True)
        mylcd.clear()
        mylcd.write_string('Printer is       Operational')
      
      if payload["state_string"] in "Cancelling":
        mylcd.clear()
        mylcd.write_string('Printer  is Cancelling job') 
        time.sleep(0.2)
      
      if payload["state_string"] in "PrintCancelled":
        mylcd.clear()
        time.sleep(0.5)
        mylcd.write_string(' Job has been Cancelled (0_0) ' ) 
        time.sleep(2)
      
      if payload["state_string"] in "Paused":
        mylcd.clear()
        time.sleep(0.5)
        mylcd.write_string('Printer is  Paused') 

      if payload["state_string"] in "Resuming":
        mylcd.clear()
        mylcd.write_string('Printer is Resuming its job') 
        time.sleep(0.2)

  def get_update_information(self):
      return dict(
          LCD1602Plugin=dict(
              displayName="LCD1602 display",
              displayVersion=self._plugin_version,

              type="github_release",
              current=self._plugin_version,
              user="n3bojs4",
              repo="OctoPrint-Lcd1602",

              pip="https://github.com/n3bojs4/octoprint-LCD1602/archive/{target}.zip"
          )
      )

__plugin_name__ = "LCD1602 I2c display"

def __plugin_load__():
	global __plugin_implementation__
	__plugin_implementation__ = LCD1602Plugin()

	global __plugin_hooks__
	__plugin_hooks__ = {
		"octoprint.plugin.softwareupdate.check_config": __plugin_implementation__.get_update_information
	}
